packages/react-router-dom/index.tsx

```diff
@@ -418,31 +418,31 @@ export const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
     },
     ref
   ) {
-    // `location` is the unaltered href we will render in the <a> tag for absolute URLs
-    let location = typeof to === "string" ? to : createPath(to);
-    let isAbsolute = /^(?:[a-z][a-z0-9+.-]*:|\/\/)/i.test(location);
-
-    // Location to use in the click handler
-    let navigationLocation = location;
     let isExternal = false;
-    if (isBrowser && isAbsolute) {
+    let absoluteHref;
+
+    if (
+      isBrowser &&
+      typeof to === "string" &&
+      /^(?:[a-z][a-z0-9+.-]*:|\/\/)/i.test(to)
+    ) {
```

Comment:
All absolute logic now _only_ happens when `to` is a string and we use `to` unaltered as we did before in `useHref` and `useLinkClickHandler` for non-absolute URLs